components/service/glean/src/test/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngineTest.kt

```diff
@@ -0,0 +1,340 @@
+                json.toString()
+            )
+        }
+    }
+
+    @Test
+    fun `timing distributions must not accumulate negative values`() {
+        // Define a timing distribution metric, which will be stored in "store1".
+        val metric = TimingDistributionMetricType(
+            disabled = false,
+            category = "telemetry",
+            lifetime = Lifetime.User,
+            name = "test_timing_distribution",
+            sendInPings = listOf("store1"),
+            timeUnit = TimeUnit.Millisecond
+        )
+
+        // Attempt to accumulate a negative sample
+        metric.accumulate(-1L)
+        // Check that nothing was recorded.
+        assertFalse("Timing distributions must not accumulate negative values",
+            metric.testHasValue())
+
+        // Make sure that the errors have been recorded
+        assertEquals("Accumulating negative values must generate an error",
+            1,
+            ErrorRecording.testGetNumRecordedErrors(metric, ErrorRecording.ErrorType.InvalidValue))
+    }
+
+    @Test
+    fun `overflow values accumulate in the last bucket`() {
+        // Define a timing distribution metric, which will be stored in "store1".
+        val metric = TimingDistributionMetricType(
+            disabled = false,
+            category = "telemetry",
+            lifetime = Lifetime.User,
+            name = "test_timing_distribution",
+            sendInPings = listOf("store1"),
+            timeUnit = TimeUnit.Millisecond
+        )
+
+        // Attempt to accumulate an overflow sample
+        metric.accumulate(TimingDistributionData.DEFAULT_RANGE_MAX + 100)
+        // Check that timing distribution was recorded.
+        assertTrue("Accumulating overflow values records data",
+            metric.testHasValue())
+
+        // Make sure that the overflow landed in the correct (last) bucket
+        val snapshot = metric.testGetValue()
+        assertEquals("Accumulating overflow values should increment last bucket",
+            1L,
+            snapshot.values["${TimingDistributionData.DEFAULT_BUCKET_COUNT - 1}"])
+    }
+
+    @Test
+    fun `getBuckets() correctly populates the buckets property`() {
+        // Hand calculated values using current default range 0 - 60000 and bucket count of 100
+        val testBuckets: List<Long> = listOf(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 17,
+            19, 21, 23, 25, 28, 31, 34, 38, 42, 46, 51, 56, 62, 68, 75, 83, 92, 101, 111, 122, 135,
+            149, 164, 181, 200, 221, 244, 269, 297, 328, 362, 399, 440, 485, 535, 590, 651, 718,
+            792, 874, 964, 1064, 1174, 1295, 1429, 1577, 1740, 1920, 2118, 2337, 2579, 2846, 3140,
+            3464, 3822, 4217, 4653, 5134, 5665, 6250, 6896, 7609, 8395, 9262, 10219, 11275, 12440,
+            13726, 15144, 16709, 18436, 20341, 22443, 24762, 27321, 30144, 33259, 36696, 40488,
+            44672, 49288, 54381, 60000, 60001)
```

Comment:
The last bucket with a width of 1 seems odd.
Could this just end with `..., 54381, 6000)` or `..., 54381, 6001)`?

Reply:
It might also just be fine to file a follow-up about the bucket fine-tuning for the next iteration.

Reply:
That last bucket is actually just the overflow bucket.  It's size of `60001` is just an arbitrary side effect of the algorithm that generates the buckets (which I copied from desktop, so I hope it's right)

Reply:
Sure, the overflow is fine. The odd part is the 1-width bucket of `6000`.

If you look at [visualizations](https://mzl.la/2YpTGoK) of histogram data, a 1-sized bucket will be hard to make sense of for analysis, it doesn't fit the exponential bucket pattern.

Reply:
I hesitate to re-design the algo that I was told was used on both the client and pipeline side (the bucket sizing algo), so @Dexterp37 any thoughts here?

Reply:
Making a [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1538750) to come back to this and discuss it more thoroughly and validate the bucket generation algorithm